### PR TITLE
[CI] [Cypress] Add E2E tests to Deploy Filters Configuration through Meshery UI

### LINKED ISSUE #4582: [CI] Add integration tests for "/configuration/*" for patterns, filters and applications
#### Current Behavior

There aren't any integration tests that cover Meshery's configuration management functionality.

#### Desired Behavior

An integration test should be written to cover these functionalities.

#### Implementation Notes

* Patterns feature is completely functional now. You should be able to perform CRUD operations with a [valid pattern file](https://github.com/service-mesh-patterns/service-mesh-patterns/tree/master/samples) and configure, apply these.
* Filters are partially functional. Basic CRUD operations are supported. [Sample filters](https://github.com/layer5io/wasm-filters).
* Applications work similar to patterns.

#### Acceptance Tests

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and Resources
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### LINKED ISSUE #4826: [CI] [Cypress] Add E2E tests to Deploy Applications/Filters/Patterns Configuration through Meshery UI
Child issue of Umbrella Issue #4582

Current Behavior
No E2E tests that cover Meshery's Service Mesh Configuration Management functionality (Applications/Filters/Patterns) to **Deploy Configuration Changes.**

Desired Behavior
E2E tests should be written to cover Deploying Applications, Filters & Patterns Configuration Changes.

Implementation Notes
Grouping E2E tests by usecase / end user scenario + multiple pages is to focus on test code reusage to try to reduce maintenance efforts & enforce standard expectations & behavior across the three pages

Acceptance Tests

---
#### Contributor [Guides](https://docs.meshery.io/project/contributing) and Resources
- 🛠 [Meshery Build & Release Strategy](https://docs.meshery.io/project/build-and-release)
- 📚 [Instructions for contributing to documentation](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#documentation-contribution-flow)
   - Meshery documentation [site](https://docs.meshery.io/) and [source](https://github.com/meshery/meshery/tree/master/docs)
- 🎨 Wireframes and designs for Meshery UI in [Figma](https://www.figma.com/file/SMP3zxOjZztdOLtgN4dS2W/Meshery-UI)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by MarioArriaga92**
**Description**

This PR fixes #4826 (partially)

No E2E tests that cover Meshery's Service Mesh Configuration Management functionality (Applications/Filters/Patterns) to Deploy Configuration Changes.

- [x] Add generic `data-cy` attributes for MesheryApplications, Filters, MesheryPatterns & Header components to enable cypress test code reuse for page titles, table grid, table grid rows, File Upload input,  & Deploy `ArrowIcon` buttons.
- [x] Add fixtures `/configuration/service-mesh-atterns-samples` folder with bookInfoPattern & IstioFilterPattern yaml files named with 8 first characters of commit hash as an attempt to "version" test data files being used.
- [x] Add fixtures `/configuration/wasm-filters-v0.1.2` folder with http-auth-envoy yaml file (folder named with `layer5io/wasm-filters` repo latest releaste name to "version" test data file being used).
- [x] Add /cypress/integration/e2e/service_mesh_configuration_management_spec.js file where "/configuration/*" E2E tests for Applications, Filters & Patterns BETA Pages (no mocks, requests made from Meshery UI served by Meshery Server itself)
- [x] Add `configurationTestTemplate` function that receives `configurationTests` array with an object per test in spec file mentioned above to reuse configuration E2E cypress tests defined in it.
- [x] Add `configurationTests` array that defines pseudo-standard testData for filter, pattern & application `itemType`
- [x] Add `Deploys ${itemType}` **test template** deploy test that opens Meshery UI through actual browser, navigates to respective page, uploads file, clicks deploy arrow button & ensures deployment is successful.
- [x] Test Data for Deploy Filter test with 'wasm-filters-v0.1.2/http-auth-envoy.yaml' 
- [x] **Commented Out** Test Data for failing Deploy Pattern test with 'service-mesh-patterns-samples/IstioFilterPattern-commit-b23b219.yaml'
- [x] **Commented Out** Test Data for failing Deploy Application test with 'service-mesh-patterns-samples/bookInfoPattern-commit-5eb2369.yaml'

**Notes for Reviewers**
1. I've added integration tests for Filters GET & POST from /api/filter endpoint, still needs some more work but I'd like to receive any feedback about overall approach before moving forward and implementing more tests.
2. Commented tests will be uncommented until necessary changes are done in necessary places in order for those tests to pass **(either Meshery Server and/or UI that causes current “invalid workload definition” and 'rpc error: code = Unknown desc = no matches for kind "EnvoyFilter" in version "networking.istio.io/v1alpha3"\n')**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ X] Yes, I signed my commits.

### CODE REVIEW

File: ui/cypress/integration/integration/configuration_spec.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+describe('Configuration Management', () => {
+  beforeEach(() => {
+    // Prepare Stub Interception for Initial Filters
+    cy.intercept('GET', '/api/filter**', { fixture : 'configuration/filter-stubs/initial-filters.json' }).as('getInitialFilters');
```
Comment (Abhishek-kumar09) [inline ui/cypress/integration/integration/configuration_spec.js:4]:
```suggestion
    cy.intercept('GET', '/api/filter**', { fixture : 'configuration/filter-stubs/initial-filters.json' }).as('getInitialFilters');
    cy.selectProviderNone();
```
Without selecting providers, we won't be able to perform operations, I think.

Comment (MarioArriaga92) [reply ui/cypress/integration/integration/configuration_spec.js:4]: @Abhishek-kumar09 thanks for suggestion, although in a prior commit adding the `cy.selectProviderNone()` custom command threw errors:
https://github.com/meshery/meshery/pull/4717/commits/288775510ef8dcd3cc6a5e0ebc822b34edf73894

Comment (Abhishek-kumar09) [reply ui/cypress/integration/integration/configuration_spec.js:4]: @MarioArriaga92 I think, I can't push some changes to your branch. How can I do that? Any idea.

Comment (Abhishek-kumar09) [reply ui/cypress/integration/integration/configuration_spec.js:4]: > @Abhishek-kumar09 thanks for suggestion, although in a prior commit adding the `cy.selectProviderNone()` custom command threw errors: [2887755](https://github.com/meshery/meshery/commit/288775510ef8dcd3cc6a5e0ebc822b34edf73894)

Oh I see, I just ran all the tests successfully after this. Maybe you can re-assure.

Comment (MarioArriaga92) [reply ui/cypress/integration/integration/configuration_spec.js:4]: @Abhishek-kumar09 let's keep in mind that tests inside `/ui/cypress/integration/integration/` folder run in .github/workflows/ci.yml file's [`tests-ui` job](https://github.com/meshery/meshery/blob/master/.github/workflows/ci.yml#L108) which only relies on `build-ui` job (`Run tests` run step only starts Meshery UI in current Github Actions runner during [`npm run ci-test-integration`](https://github.com/meshery/meshery/blob/master/.github/workflows/ci.yml#L136)).

Here's failure that occured when suggested custom command was used in `beforeEach` hook:
https://github.com/meshery/meshery/runs/4412890391?check_suite_focus=true

Comment (MarioArriaga92) [reply ui/cypress/integration/integration/configuration_spec.js:4]: Current `Meshery / UI integration tests (pull_request)` check in PR footer shows success on latest run and drilling into its 'Details' would bring up run output:
Screen Shot 2021-12-09 at 9 43 05

Comment (MarioArriaga92) [reply ui/cypress/integration/integration/configuration_spec.js:4]: @Abhishek-kumar09 Tests run successfully with `npm run ci-test-integration` right from /ui/ folder so resolving this conversation, but let's keep conversation going about simplifying working with cypress locally right on slack 🔥👌🙏🏼

Comment (Abhishek-kumar09) [reply ui/cypress/integration/integration/configuration_spec.js:4]: Sure :muscle:

**Comment by Abhishek-kumar09** [comment]
@leecalcote @ramrodo @MarioArriaga92  Considering the cypress intergration tests, should it run without server?
because server is one that loads capabilities, actually uploads files and perform operations. Although we can mock server but the UI is highly dependent on the server for all the things it shows. e.g. Patterns/filters/applications should not open without logging in with providers/none-provider. 

currently, We are able to open patterns/filters/application endpoint in the UI without the servers, that is a bug?

**Comment by MarioArriaga92** [comment]
@Abhishek-kumar09 I think we can create a branch in Meshery/Meshery repo so multiple contributers can work on related changes, can you guys confirm this is the usual approach? @leecalcote @navendu-pottekkat @utkarsh-pro @DelusionalOptimist?

**Comment by MarioArriaga92** [comment]
@Abhishek-kumar09  @leecalcote @navendu-pottekkat @utkarsh-pro @DelusionalOptimist Reason for above suggestion is that while #4582's scope is just for "/configuration" frontend route the actual test scope (Applications, Patterns & Filters) might be as big enough for two or more contributors to collaborate on and multiple PRs might make sense in order to tackle it. Especially if both "integration" and "e2e" tests are to be considered for each page along with CRUD for each use case

**Comment by MarioArriaga92** [comment]
Actually, let me discuss with @Abhishek-kumar09 about cypress effort strategy. We should be able to reduce scenarios to just the most important ones but need to go over testing spreadsheet & some brainstorming

**Comment by navendu-pottekkat** [comment]
@MarioArriaga92 Creating an "epic" issue with a bunch of "child" issues under it would be the way to go.

Merging each PR can close the child issues and the epic issue can stay open. This makes it easy to track progress of each area as well as the overall goal of writing integration tests.

We generally don't branch out unless it is absolutely necessary. And since these integration tests wouldn't affect any functionality, we can keep merging to master.

**Comment by ramrodo** [comment]
@MarioArriaga92 Great work! No comments so far. You're mocking the responses well and describing/adding good comments in some tests. The GET & POST tests look good to me.

**Comment by ramrodo** [comment]
> @leecalcote @ramrodo @MarioArriaga92 Considering the cypress intergration tests, should it run without server? because server is one that loads capabilities, actually uploads files and perform operations. Although we can mock server but the UI is highly dependent on the server for all the things it shows. e.g. Patterns/filters/applications should not open without logging in with providers/none-provider.

@Abhishek-kumar09 thanks for pointing it out but the Integration tests should be run without running the backend/server. If we want to test both and the whole function as a user, we need to create E2E tests. This article of [Static vs Unit vs Integration vs E2E Testing for Frontend Apps](https://kentcdodds.com/blog/static-vs-unit-vs-integration-vs-e2e-tests) explains the different types of tests.

**Comment by Abhishek-kumar09** [comment]
@MarioArriaga92 Probably we can merge this PR and get things done in parts. The changes and tests looks good and maybe you can open it for merging?

**Comment by MarioArriaga92** [comment]
@leecalcote @ramrodo @Abhishek-kumar09 @navendu-pottekkat 

In order to account for recent discussions & general decision to choose a more broad e2e testing approach, I've repurposed this PR **(Just updated the Description)** to solve new issue opened (#4826 ) which scopes Deploying Configuration changes on all 3 pages.

Approach for test code in /cypress/integration/e2e/service_mesh_configuration_management_spec.js is basically create a testTemplate function & pass configurationTests array in which each object represents an area to test. Further tests can/should be added inside this `testTemplate` function defined in this file

Subsequent child issues based on umbrella issue #4582 will be created for Editing, Deleting & any other critical E2E workflow (upload file > Edit > Deploy scenario, etc.) as we agree its fit.

Still working on fixing some 500's being thrown on both Applications & Patterns as well as adding a couple of missing assertions against Meshery UI to check deployment is successful **(besides checking POST /api/{itemType}/deploy endpoint responses are '200' which is already scripted)**

**Comment by l5io** [comment]
🚀 Preview for commit 14e21a5a021b875c683ddae027470093b8eaa7ae at: https://61c60a96b5f4a27f946c7d5c--meshery-docs-preview.netlify.app

**Comment by leecalcote** [comment]
Uh-oh, you might need to sync your local fork, so that only files you've changed are included in this PR. Instructions on syncing your fork - https://github.com/meshery/meshery/blob/master/CONTRIBUTING-gitflow.md

**Comment by l5io** [comment]
🚀 Preview for commit b6600deb475d3ad1471d5b45fa5f227f0225133c at: https://61c61380fd45e08cfff68766--meshery-docs-preview.netlify.app

**Comment by MarioArriaga92** [comment]
@leecalcote I did follow steps to fetch upstream, merge upstream/master to fork's master & rebased ci/marrioarriaga92/4582 branch onto master.. multiple times between pushing & pulling from local lol

**Comment by MarioArriaga92** [comment]
I'll look at instructions more closely np @leecalcote

**Comment by l5io** [comment]
🚀 Preview for commit 0f2eae4a92fa108c3f3029e4c3d7962e51fc6cc9 at: https://61c62420f3060d954228c6a6--meshery-docs-preview.netlify.app

**Comment by MarioArriaga92** [comment]
I had accidentally did a merge instead of rebase so I had to push --force to undo merge

**Comment by Abhishek-kumar09** [comment]
Pattern deploy has the problems...

**Comment by Abhishek-kumar09** [comment]
Try the simpler pattern like an application pattern.

**Comment by leecalcote** [comment]
@MarioArriaga92 any chance that you might merge the completed tests, so that the project can start benefiting now without awaiting finalization of all tests?

**Comment by warunicorn19** [comment]
beat me to it @leecalcote

**Comment by leecalcote** [comment]
30 seconds is an impressive margin all considering. 🙂

**Comment by MarioArriaga92** [comment]
Thanks @leecalcote, @warunicorn19 wasn't aware that "warunicorn19" and "Aditya Chatterjee" was the same person lol

**Pull request merged**